src/dd/dd.c

```diff
@@ -0,0 +1,441 @@
+	     templateIndex < dd->template_cnt;
+	     templateIndex++) {
+
+		for (uint8_t chainIndex = 0;
+		     chainIndex < dd->chain_cnt;
+		     chainIndex++) {
+
+			if (getbit_get_left(gb) < 4)
+				return EBADMSG;
+
+			dd->template_chain_fdiff[templateIndex][chainIndex] =
+				dd_f(4);
+		}
+	}
+
+	return 0;
+}
+
+
+static int render_resolutions(struct dd *dd, struct getbit *gb)
+{
+	for (uint8_t spatial_id = 0;
+	     spatial_id <= dd->max_spatial_id;
+	     spatial_id++) {
+
+		if (getbit_get_left(gb) < 32)
+			return EBADMSG;
+
+		dd->max_render_width_minus_1[spatial_id]  = dd_f(16);
+		dd->max_render_height_minus_1[spatial_id] = dd_f(16);
+
+		++dd->render_count;
+	}
+
+	return 0;
+}
+
+
+static int template_dependency_structure(struct dd *dd, struct getbit *gb)
+{
+	if (getbit_get_left(gb) < 11)
+		return EBADMSG;
+
+	dd->template_id_offset = dd_f(6);
+	uint8_t dt_cnt_minus_one = dd_f(5);
+
+	dd->dt_cnt = dt_cnt_minus_one + 1;
+
+	int err = template_layers(dd, gb);
+	if (err)
+		return err;
+
+	err = template_dtis(dd, gb);
+	if (err)
+		return err;
+
+	err = template_fdiffs(dd, gb);
+	if (err)
+		return err;
+
+	template_chains(dd, gb);
+
+	/* TODO  decode_target_layers() */
+
+	if (getbit_get_left(gb) < 1)
+		return EBADMSG;
+
+	dd->resolutions_present_flag = dd_f(1);
+	if (dd->resolutions_present_flag) {
+		err = render_resolutions(dd, gb);
+		if (err)
+			return err;
+	}
+
+	return 0;
+}
+
+
+static int extended_descriptor_fields(struct dd *dd, struct getbit *gb)
+{
+	re_printf("extended_descriptor_fields\n");
+
+	if (getbit_get_left(gb) < 5)
+		return EBADMSG;
+
+	dd->template_dependency_structure_present_flag = dd_f(1);
+	dd->active_decode_targets_present_flag = dd_f(1);
+	dd->custom_dtis_flag   = dd_f(1);
+	dd->custom_fdiffs_flag = dd_f(1);
+	dd->custom_chains_flag = dd_f(1);
+
+	if (dd->template_dependency_structure_present_flag) {
+
+		int err = template_dependency_structure(dd, gb);
+		if (err)
+			return err;
+
+		dd->active_decode_targets_bitmask = (1u << dd->dt_cnt) - 1;
+	}
+
+	if (dd->active_decode_targets_present_flag) {
+
+		dd->active_decode_targets_bitmask = dd_f(dd->dt_cnt);
+	}
+
+	return 0;
+}
+
+
+static void no_extended_descriptor_fields(struct dd *dd)
+{
+	dd->custom_dtis_flag   = 0;
+	dd->custom_fdiffs_flag = 0;
+	dd->custom_chains_flag = 0;
+}
+
+
+int dd_decode(struct dd *dd, const uint8_t *buf, size_t sz)
+{
+	if (!dd || !buf)
+		return EINVAL;
+
+	memset(dd, 0, sizeof(*dd));
+
+	struct getbit gb;
+
+	getbit_init(&gb, buf, sz*8);
+
+	int err = mandatory_descriptor_fields(dd, &gb);
+	if (err)
+		return err;
+
+	if (sz > 3) {
+		err = extended_descriptor_fields(dd, &gb);
+		if (err)
+			return err;
+
+		dd->ext = true;
+	}
+	else {
+		no_extended_descriptor_fields(dd);
+	}
+
+#if 0
+	frame_dependency_definition();
+	zero_padding = f(sz * 8 - TotalConsumedBits);
+#endif
+
+	return 0;
+}
+
+
+void dd_print(const struct dd *dd)
+{
+	if (!dd)
+		return;
+
+	printf("~~~~ DD: ~~~~\n");
```

Comment:
Is there a specific reason for using `printf` instead of `re_printf` in this function?

Reply:
Thanks for the review.

I changed the code to `re_printf`